### ISSUE TITLE
cluster

### ISSUE THREAD

**Issue by morganherlocker** [Turfjs/turf#33]
Point clustering for rendering lots of points

**Comment by morganherlocker** [comment]
Ok, an O(n**2-ish) option that might be a good standin because I already wrote it :stuck_out_tongue_winking_eye: 

```js
var featurecollection = require('turf-featurecollection');
var distance = require('turf-distance');

module.exports = function(pts, maxDistance, units) {
  var clusters = turf.featurecollection([]);
  for(var i = 0; i < pts.features.length; i++) {
    var exemplar = true;
    for(var k = 0; k < clusters.features.length; k++){
      if(turf.distance(pts.features[i], clusters.features[k], units) < maxDistance) {
       exemplar = false;
     }
    }
    if(exemplar) clusters.features.push(pts.features[i]);
  }

  return clusters;
}
```

A slower and more memory hungry version would accumulate the cluster points and compute the centroid at the end. I think the tradeoffs might be worthwhile, but I'm going to stew on this for a bit.

**Comment by stepankuzmin** [comment]
Hi @morganherlocker! Is there are any progress with this?

**Comment by rowanwins** [comment]
Hi @stepankuzmin it looks like this has largely been tackled with the [supercluster ](https://github.com/mapbox/supercluster) library from mapbox. Documentation is a little sparse but hopefully if you follow the tests you'll get a few ideas.

Cheers
Rowan

**Comment by cyrilchapon** [comment]
Clustering on a map and clustering some points is a very different feature.

I personaly need K-mean clustering inside an API that serves geojson.

So, is there any progress with this ?

**Comment by morganherlocker** [comment]
> Clustering on a map and clustering some points is a very different feature.
> I personaly need K-mean clustering inside an API that serves geojson.
> So, is there any progress with this ?

Agree. No progress to report, but I still think this would be a great feature.

**Comment by donmccurdy** [comment]
Would this differ from what 'supercluster' provides? Note that it can be used serverside, in which case it takes geojson points as input, and outputs geojson points for the clusters, annotated with point counts.

**Comment by cyrilchapon** [comment]
Very thanks on documentation.

Well the zoomlevel thing config is philosohically anoying, but I should be able to deal with it 😄

I'm trying hard, I'll tell you

**Comment by cyrilchapon** [comment]
I'm just struggling at making it work.

It aggregates all my points to a single point.

**Comment by morganherlocker** [comment]
> Would this differ from what 'supercluster' provides? 
> [...]
> Well the zoomlevel thing config is philosohically anoying, but I should be able to deal with it 😄

supercluster is geared towards rendering, from an interface perspective and performance perspective. Turf tends to take a more generic statistical approach. I would recommend using one of the existing kmeans implementations in node for now, or using supercluster if it fits your use case (rendering or perhaps multilevel quadtree clustering). If you have any more questions regarding supercluster, please open an issue on its repo.

**Comment by cyrilchapon** [comment]
@morganherlocker I basicaly agree with you.

Though I posted here originaly because I believe there is a need **beyond** existing *k-means implementations* (simply aggregating) but not as far as *supercluster* goes (rendering oriented).

I think there is a need, in between, to simply aggregate GeoJSON point. *Turf* is a wonderful library that deals with geospatial manipulation and GeoJSON. Therefore I think turf would be a right place for something fitting this need.

**Comment by morganherlocker** [comment]
Yep, just to clarify the [original idea here](https://github.com/Turfjs/turf/issues/33#issuecomment-28069237). This feature would ideally group points into `k` clusters based on longitude and latitude. It would be independent of rendering, and could be used for grouping and aggregate statistics.

An API would look something like this:

```js
var cluster = require('@turf/cluster')

var pts = // featurecollection of points
var numGroups = 10

var clustered = cluster(pts, numGroups)

// clustered is a new featurecollection of points with the
// following new properties for cluster groups:
// - clusterID
// - cluster exemplar: a geojson point representing the cluster's centroid
```

**Comment by cyrilchapon** [comment]
- In the process it could accept some kind of `aggregator` function. The next step beyond clustering indeed seems to be aggregating in the process (at least IMHO and in my current use case), I've found issues (supercluster) and SO questions related to this so. We could aggregate inside each cluster (represented by a Feature)' properties, maybe.

- It also should include a `radius`.

- Also, what is `numgroups` ?

- Finally, I think it could return a `FeatureCollection<Feature<GeometryCollection>>` rather than a `FeatureCollection<Feature<Point>>`. The GeometryCollection then could have indeed the **centroid** but also the **outer polygon**, maybe ? Here's the spec as I'd see it.

```
var cluster = require('@turf/cluster')

var pts = // FeatureCollection<Feature<Point>> (or array<Feature<Point>> ?)
var options = {
  numGroups: 10,
  radius: 0.15,
  radiusUnit: 'kilometers',
  aggregator: function(propertiesAccumulator, featureProperties) {
    //Example val1 is a average, val2 is a sum and val3 is a min
    return _.extend({}, propertiesAccumulator, {
       count: propertiesAccumulator.count + 1,
       val1: propertiesAccumulator.val1 ? ((propertiesAccumulator.val1 * propertiesAccumulator.count) + featureProperties.val1) / propertiesAccumulator.count + 1 : featureProperties.val1,
       val2: propertiesAccumulator.val2 ? propertiesAccumulator.val2 + featureProperties.val2 : featureProperties.val2,
       val3: !propertiesAccumulator.val3 || featureProperties.val3 < propertiesAccumulator.val3 ? featureProperties.val3 : propertiesAccumulator.val3
    })
  },
  initialAccumulator: {
    count: 0,
    val1: null,
    val2: null,
    val3: null
  }
}

var clustered = cluster(pts, options)
```

clustered is a new `FeatureCollection<Feature<GeometryCollection>>` with the following:

```
// {
//   type: 'FeatureCollection',
//   features: [
//     type: 'Feature',
//     properties: {
//       clusterId: ???,
//       val1: 14,
//       val2: 36,
//       val3: 47
//     },
//     geometry: {
//       type: 'GeometryCollection',
//       geometries: [
//         {
//           type: 'Point',
//           coordinates: [3, 50]
//         },
//         {
//           type: 'Polygon',
//           coordinates: [[[3, 50], [4, 50], [4, 60], [3, 60], [3, 50]]]
//         }
//       ]
//     }
//   ]
// }
```

**Comment by morganherlocker** [comment]
Thanks for the notes @cyrilchapon !

> It also should include a radius.
> Also, what is numgroups ?

I am proposing that this module implements spatial [k-means clustering](https://en.wikipedia.org/wiki/K-means_clustering). `numGroups` is the `k` variable which determines the number of clusters that will be generated. The "radius" of a cluster would be dynamic based on what the algorithm determines is the best way to split the boundaries, but its important to note that clusters may have overlapping bounding circles. It sounds like this may be a different type of module than what you are looking for.

> In the process it could accept some kind of aggregator function. 

My preference is to leave aggregation method to the user, since JavaScript provides builtin functions for this sort of thing that work great. Typically a turf function takes a static input and returns a static output. I believe this greatly simplifies the interfaces between modules and leaves the door open for declarative interfaces.

> The GeometryCollection then could have indeed the centroid but also the outer polygon, maybe?

These clusters do not have a defined outer polygon. It is a subjective decision as to what defines the hull of the cluster, so it should be left to the user. A few options for generating the hull include: calculating minimum bounding circle, dissolved voronoi cells, turf-bbox, turf-concave, and turf-convex.

#### misc

- GeometryCollection could be a good option here. My one concern, and the reason I suggested a denormalized format, was that a GeometryCollection makes it difficult to iterate over the properties of each point within a cluster. This is a common operation for people crunching stats on a set of clusters. Another option could be a FeatureCollection<Feature<Point>>> where each Point is a centroid with an `indices` property that is an array of indices to the original FeatureCollection's points. This would avoid duplicating data.

- Instead of an `options` object, the general convention is to enumerate each option as an explicit parameter. This makes it easier to set expectations about what a user needs to understand/decide and it helps with automatically generating documentation.

**Comment by tmcw** [comment]
Another option for the geometry output:

- Output the input, with each feature given a `cluster` property for its cluster #. And output, as a separate featurecollection, the cluster centroids also with their numbers.

**Comment by cyrilchapon** [comment]
@morganherlocker, thanks for your answer.
> I am proposing that this module implements spatial k-means clustering.
> It sounds like this may be a different type of module than what you are looking for.

Hum, yes actualy, I had misunderstood. Though... I'm posting it here because this is "cluster" topic issue, and I'm looking for "clusters". Just not k-means one I guess (not a Math guy though).

> My preference is to leave aggregation method to the user

Yes this is true. Though, in consequence, the output should be in a form that let us aggregate manualy after this. The only solution, I think, is the tmcw one :

> Output the input, with each feature given a cluster property for its cluster #. And output, as a separate featurecollection, the cluster centroids also with their numbers.

------

I agree about miscs

**Comment by stebogit** [comment]
Implemented with #787
